File: CS50_CS/Problems/Week_6/credit.py
def main():
    number = input("Number: ")
    n = str(len(number))
    cards = ['13', '15', '16']
    master=['51','52','53','54','55']
    if n not in cards:
        return print("INVALID")
    if (number[0:2] == '34' or number[0:2] == '37') and validate(number) == True:
        return print("AMEX")
    elif number[0:2] in master and validate(number) == True:
        return print("MASTERCARD")
    elif number[0] == '4' and validate(number) == True:
        return print("VISA")
    else:
        return print("INVALID")


def validate(number):
    number=number[::-1]
    digits = []
    counter = 1
    for digit in number:
        if counter % 2 == 0:
            if int(digit)*2 >= 10:
                second = int(digit)*2 % 10
                first = 1
                digits.append(first)
                digits.append(second)
            else:
                digits.append(int(digit)*2)
        counter += 1
    odds = []
    counter = 1
    for digit in number:
        if counter % 2 == 1:
            odds.append(int(digit))
        counter += 1
    final = sum(digits)+sum(odds)
    if final % 10 == 0:
        return True
    else:
        return False


if __name__ == "__main__":
    main()

File: CS50_CS/Problems/Week_1/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    long number = get_long("Number: ");
    int n = floor(log10(number)) + 1;
    int arr[n] = {};
    int i = 0;
    while (number != 0)
    {
        int digit = number % 10;
        number = number / 10;
        arr[i] = digit;
        i++;
    }
    string type;
    if (n == 15 && arr[n - 1] == 3 && (arr[n - 2] == 4 || arr[n - 2] == 7))
    {
        type = "AMEX";
    }
    else if (n == 16 && arr[n - 1] == 5 && arr[n - 2] >= 1 && arr[n - 2] <= 5)
    {
        type = "MASTERCARD";
    }
    else if ((n == 13 || n == 16) && arr[n - 1] == 4)
    {
        type = "VISA";
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
    int factor;
    int first;
    int second;
    int sum = 0;
    if (n % 2 == 0)
    {
        for (int j = n - 1; j >= 1; j -= 2)
        {
            factor = arr[j] * 2;
            if (factor > 9)
            {
                second = factor % 10;
                first = factor / 10;
                sum += first;
                sum += second;
            }
            else
            {
                sum += factor;
            }
        }
        for (int k = 0; k <= n - 1; k += 2)
        {
            sum += arr[k];
        }
        if (sum % 10 == 0)
        {
            printf("%s\n", type);
            return 0;
        }
        else
        {
            printf("INVALID\n");
            return 0;
        }
    }
    else
    {
        for (int j = n - 2; j >= 1; j -= 2)
        {
            factor = arr[j] * 2;
            if (factor > 9)
            {
                second = factor % 10;
                first = factor / 10;
                sum += first;
                sum += second;
            }
            else
            {
                sum += factor;
            }
        }
        for (int k = 0; k <= n - 1; k += 2)
        {
            sum += arr[k];
        }
        if (sum % 10 == 0)
        {
            printf("%s\n", type);
            return 0;
        }
        else
        {
            printf("INVALID\n");
            return 0;
        }
    }
}
